src/lib/core/exception.h

```diff
@@ -257,6 +257,22 @@ class RaftError: public Exception {
 	virtual void raise() { throw this; }
 };
 
+class HandleIsClosed: public Exception {
+public:
+	HandleIsClosed(const char *file, unsigned line,
+		       const char *format, ...);
```

Comment:
`ChannelIsClosed` doesn't have the `format` argument. Why does `HandleIsClosed` need it?

Also, maybe we should pass some extra information to the error? For example what kind of handle is closed.

Also, do we really need a generic error for this kind of thing? Maybe, add `PopenHandleIsClosed` error specifically for popen? I'd prefer a specific error - it's unclear if we ever want to reuse `HandleIsClosed` for other kinds of handles.

Reply:
> `ChannelIsClosed` doesn't have the `format` argument. Why does `HandleIsClosed` need it?

If it would be `PopenHandleIsClosed` I wouldn't add the error message. The primary reason to add it is to say what kind of a handle is closed.

> Also, maybe we should pass some extra information to the error? For example what kind of handle is closed.

A separate field? I doubt that users ever know existing fields: we have [almost no documentation](https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_error/error/) about the box error API. I think that `err.type` + `err.message` (where a kind of handle will be mentioned) should be enough.

> Also, do we really need a generic error for this kind of thing? Maybe, add `PopenHandleIsClosed` error specifically for popen? I'd prefer a specific error - it's unclear if we ever want to reuse `HandleIsClosed` for other kinds of handles.

I have no strict objections against the specific error, but I would prefer to add a general error that would be usable in other similar situations.

(I would also add that I don't like `<Subsystem>Error`: if you ever want to differentiate exit states of a function, all its errors will be from the same subsystem. You likely will need to make a decision like retry/log and ignore/wrap and report upward and subsystem doesn't help here anyhow. *What* occurs -- it should be a first level of an error hierarchy, as for me.)

We discussed it with Vladimir and he also highlighted possible point of confusion in my approach: if a handle is named as descriptor for some object kind, the error name may confuse someone.

We also discussed, whether the error should be returned (as `nil, err`) or raised. There are cons and pros -- we didn't end with a final agreement here.

* For `nil, err`: now we have the simple 'raise IllegalParams and FiberIscancelled, return everything else' rule.
* For `error(err)`: the error is a kind of 'real error' -- something external affect our code and we can't do much here. Vladimir proposes to raise such errors.

OTOH, we have SystemError returned, while it falls into the same 'real error' category.

Reply:
> OTOH, we have SystemError returned, while it falls into the same 'real error' category.

SystemError is a 'normal' error - it may happen during normal program operation. For example, a program should handle situations like a missing file or a not responding host.

Just to be clear, I'd prefer to add a new error type without any parameters for this situation - `PopenHandleIsClosed` - and raise this error to Lua if the handle is closed from under us.

BTW, looks like we do raise an error on an attempt to use a closed handle:

https://github.com/tarantool/tarantool/blob/c9155ac8636264ce6d1b10a2937a6ad3844db50a/src/lua/popen.c#L597-L602

Currently, we use `IllegalParams` error, but I think we should use `PopenHandleIsClosed` here, as well.

Reply:
> BTW, looks like we do raise an error on an attempt to use a closed handle:

Sure. When it is known as closed at the moment of the call and the call is definitely incorrect (a user can validate it in their code). I handle it similarly to any other incorrect parameter.

Whether we should report closed-while-the-call-in-progress situation in the same way or differently -- is the open question.

> Currently, we use `IllegalParams` error, but I think we should use `PopenHandleIsClosed` here, as well.

We already use `IllegalParams` here, I can't change the API.

Reply:
> We already use `IllegalParams` here, I can't change the API.

I think you can.

Reply:
It would be more or less okay (not sure, though) if I wouldn't add all exit situations with their error types into the documentation.

As a general rule, when 'make it consistent/beautiful' and 'keep it compatible' fight, the latter should win.

Reply:
Error messages are subject to change - it's perfectly normal (think of localization). A user program that checks error messages is broken by design. Error codes shouldn't change (ideally), but AFAIU nobody checks error codes for errors like this in Lua (?). At least, we often update error codes from version to version (e.g. reuse dropped error codes). Since there's no common mechanism for catching errors by type in Lua (AFAIK), we can change error codes, as well. Anyway, `PopenHandleIsClosed` looks like a more apt error for an attempt to use a closed handle than `IllegalParams`.

Reply:
`err.type` will give `'IllegalParams'`, `'TimedOut'`, `'SystemError'` and so on.

Well, we have really stochastic error system. There are two ways to going toward a strict API/good error reporting system:

* Revolutionary way. Don't take care to errors as we usually do now. In a future (which will never occur) design a brand new error reporting system and start to follow the new rules. Of course, this design will be more theoretical than practical and has high probability to don't actually solve practical problems.
* Evolutionary way. Attempt to start doing the right things now: take care to compatibility, take care to proper naming/hierarchy. Maybe we'll anyway end with necesarity to proceed with the revolutionary way, but the new design will be based on the practical problems we'll meet on the road.

popen is the module, where I'm trying to go in the second way. I can (but don't want) accept 'it does not work anyway' point.

Reply:
I still don't like the idea of using `Popen<...>` error class in one specific case, but use general purpose error classes in all other diagnostics of the module.

I'm going to just use `SystemError` and zap the `strerror()` postfix.

I see that we use it as a general error class in `tuple_convert.c`, `httpc.c`. However, it doesn't look as a widespread practice.

Is it acceptable?

Reply:
It doesn't look like a `SystemError` to me, and it doesn't have `errno`. Let's use `ChannelIsClosed`.

Reply:
OK. Changed the error type to `ChannelIsClosed`.